### PR TITLE
implements the #75 reader writer lock

### PR DESCRIPTION
**Pull request by hoswey**
Please help to review,  it seems the test case is hard to write for such concurrent pattern. :(

### CODE REVIEW

**Comment by iluwatar** [comment]
Review comments:

- The application does not exit correctly when `App.java` is run from IDE
- The example does not have proper unit tests
- The explanation of the pattern in `App.java` is very brief. Could the explanation be expanded a bit?

File: reader-writer-lock/index.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+layout: pattern
+title: Producer Consumer
```
Comment (iluwatar) [inline reader-writer-lock/index.md]: The title is wrong

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/App.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+   */
+  public static void main(String[] args) {
+
+    ExecutorService es = Executors.newFixedThreadPool(1000);
+    ReaderWriterLock lock = new ReaderWriterLock();
+
+    AtomicInteger index = new AtomicInteger(0);
+    IntStream.range(0, 100).forEach(i -> {
+      Runnable task = null;
+      if (ran.nextFloat() <= 0.6) {
```
Comment (iluwatar) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/App.java]: Why use random numbers here? I think you can just allocate certain number of readers and writers.

**Comment by iluwatar** [comment]
@hoswey you have my review remarks. Please comment on this thread once ready for another look.

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+          if (isLockFree()) {
+            g.add(this);
+            break;
+          } else if (isWriterOwnThisLock()) {
+            waitUninterruptely(g);
+          } else if (isReaderOwnThisLock()) {
+            waitUninterruptely(g);
+          } else {
+            throw new RuntimeException("it should never reach here");
```
Comment (iluwatar) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: You should not throw raw `RuntimeException`s but subclasses of this. See [AvoidThrowingRawExceptionTypes](http://pmd.sourceforge.net/pmd-4.3.0/rules/strictexception.html)

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    public void unlock() {
+
+      synchronized (g) {
+        g.remove(this);
+        g.notifyAll();
+      }
+    }
+  }
+
+  private boolean isWriterOwnThisLock() {
```
Comment (npathai) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: Better naming convention would be `doesWriterOwnThisLock()` and `doesReaderOwnThisLock()`

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    return g.contains(readLock);
+  }
+
+  private boolean isLockFree() {
+    return g.isEmpty();
+  }
+
+
+
+  private void waitUninterruptely(Object o) {
```
Comment (npathai) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: Typo: `waitUninterruptibly()`

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+/**
+ * Class responsible for control the access for reader or writer
+ */
+public class ReaderWriterLock {
+
+  /**
+   * Mutex for reader
+   */
+  private Object r = new Object();
```
Comment (npathai) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: single character names should not be promoted, I would propose `mutex` or `readMutex` something in that line which also describes what it is about. And then the javadoc on private variable will not be needed.

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
+package com.iluwatar.reader.writer.lock;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Class responsible for control the access for reader or writer
```
Comment (npathai) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: Javadoc can be a bit more detailed here. "Allows multiple readers to hold the lock at same time, but if any writer holds the lock then readers wait. If reader holds the lock then writer waits." This lock is _not fair_.

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+  /**
+   * Global mutex for reader or writer, use to save the holding object
+   */
+  private Set<Object> g = new HashSet<>();
+
+  /**
+   * Current reader count
+   */
+  private int readerCount = 0;
```
Comment (npathai) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: rename the private variable to `currentReaderCount` and remove the javadoc comment.

File: reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java (hunk 3 above)
Comment (npathai) [inline reader-writer-lock/src/main/java/com/iluwatar/reader/writer/lock/ReaderWriterLock.java]: A better approach is to throw `AssertionError`. It should be used in such cases.

**Comment by hoswey** [comment]
Thanks @iluwatar  and @npathai 's comments, the fix is done, please help to review. Thanks.

**Comment by iluwatar** [comment]
Great work with this pattern and the tests. Thanks @hoswey once again :+1:

**Comment by iluwatar** [comment]
@hoswey it seems that the unit tests are failing on Travis builds. Can you please check issue https://github.com/iluwatar/java-design-patterns/issues/343

**Comment by npathai** [comment]
Good work @hoswey . I will have to go through the logic once more as you already stated that concurrency patterns are quite complex and there can be subtle issues lurking. If there is something I will let you know and create an issue for the same.

**Pull request merged**